qutebrowser/utils/psl.py

```diff
@@ -0,0 +1,145 @@
+
+    _rules = None
+
+    def _load_rules(self) -> None:
+        PSL_PATHS = (
+            # User-provided list in data directory (need to wait until standarddir is
+            # initialized)
+            os.path.join(standarddir.data(), 'public_suffix_list.dat'),
+            # Bundled list
+            os.path.join(__file__, '../public_suffix_list.dat'),
+            # Debian publicsuffix package
+            '/usr/share/publicsuffix/public_suffix_list.dat'
+        )
+        path = None
+        for psl_path in PSL_PATHS:
+            if os.path.exists(psl_path):
+                path = psl_path
+                break
+        if not path:
+            raise PSLNotFound(PSL_PATHS)
+
+        self._rules = {}
+        with open(path, 'r') as f:
+            for line in f:
+                line = line.strip()
+                if line.startswith('//') or not line:
+                    continue
+                if line.startswith('*.'):
+                    self._rules[line[2:]] = _Rule.WILDCARD
+                elif line.startswith('!'):
+                    self._rules[line[1:]] = _Rule.EXCEPTION
+                else:
+                    assert '*' not in line and '!' not in line, line
+                    self._rules[line] = _Rule.NORMAL
+
+    @property
+    def rules(self):
+        if self._rules is None:
+            self._load_rules()
+        return self._rules
+
+    def sld(self, url: QUrl) -> str:
+        urlutils.ensure_valid(url)
+        host = url.host()
+
+        if '.' not in host or self.rules.get(host) in (_Rule.NORMAL, _Rule.WILDCARD):
+            return None  # FIXME why?
+            # If the hostname doesn't have a dot it's either a tld, which doesn't have
+            # a sld, or it doesn't match any rule, in which case the prevailing rule is
+            # '*' according to https://publicsuffix.org/list/. If the hostname has a dot
+            # but matches a normal/wildcard rule, it's a tld.
+
+        parts = list(_widened_hostnames(host))
+
+        for i, part in enumerate(parts):
+            rule = self.rules.get(part)
+            if rule == _Rule.EXCEPTION:
+                return parts[i]
+            elif rule == _Rule.NORMAL:
+                return parts[i - 1]
+            elif rule == _Rule.WILDCARD:
+                return parts[i - 2] if i >= 2 else None
+
+        return parts[-2]
+
+    def same_domain(self, url1: QUrl, url2: QUrl) -> bool:
+        """Check if url1 and url2 belong to the same website.
+
+        If the URL's schemes or ports are different, they are always treated as not
+        equal.
+
+        Return:
+            True if the domains are the same, False otherwise.
+        """
+        if url1.scheme() != url2.scheme() or url1.port() != url2.port():
+            return False
+
+        sld1 = self.sld(url1)
+        sld2 = self.sld(url2)
+        if not sld1:
+            return url1.host() == url2.host()
+
+        return sld1 == sld2
+
+
+suffix_list = _SuffixList()
```

Comment:
We don't want to run the parsing of the file more than once.

Either the module exposes this single instance instead of the class, or the class is exposed but the rules are cached in a module's variable instead of a class attribute.

Reply:
On second thought, maybe this shouldn't even be a class. A singleton which won't be treated as an object anywhere, only as a namespace for its methods, can be better represented as normal functions.

Reply:
I think having it as a class (and a singleton instance like this) is preferable - it'll also make it easier to e.g. write tests for it.